consistent-sampling/src/main/java/io/opentelemetry/contrib/samplers/ConsistentSampler.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.contrib.samplers;
+
+import static java.util.Objects.requireNonNull;
+
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.SpanKind;
+import io.opentelemetry.api.trace.TraceState;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.contrib.state.OtelTraceState;
+import io.opentelemetry.contrib.util.DefaultRandomGenerator;
+import io.opentelemetry.contrib.util.RandomGenerator;
+import io.opentelemetry.sdk.trace.data.LinkData;
+import io.opentelemetry.sdk.trace.samplers.Sampler;
+import io.opentelemetry.sdk.trace.samplers.SamplingDecision;
+import io.opentelemetry.sdk.trace.samplers.SamplingResult;
+import java.util.List;
+
+/** Abstract base class for consistent samplers. */
+abstract class ConsistentSampler implements Sampler {
```

Comment:
We want to reduce public API as much as possible in libraries. We can probably make it so this is the only public class, which has factories returning implementations, and nothing else would need to be public from what I can tell.

Reply:
Also favor composition over inheritance, so we should have a class that implements `Sampler`, containing a field that's a `ConsistentSampler` rather than talking through hierarchy

Reply:
> We want to reduce public API as much as possible in libraries. We can probably make it so this is the only public class, which has factories returning implementations, and nothing else would need to be public from what I can tell.

addressed in f7911c35c4c502b205493b02e10e5e441d47a00b

Reply:
> Also favor composition over inheritance, so we should have a class that implements `Sampler`, containing a field that's a `ConsistentSampler` rather than talking through hierarchy

I think inheritance is appropriate in this case. First, `ConsistentSampler` is a `Sampler`. And secondly, the construction of a composed consistent sampler that is the AND/OR combination of 2 consistent samplers would be more difficult otherwise.